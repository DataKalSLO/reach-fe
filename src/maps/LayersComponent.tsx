// Multiple Value Autocomplete: https://material-ui.com/components/autocomplete/
// Used for selection box, shows chip based on search and allows user to delete
import { Box } from '@material-ui/core';
import { styled } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import React from 'react';
import { flatten } from 'lodash';
import kitchenFaciltiesHeatMap from '../common/assets/Local Data/census/b25053.js';
import medianHouseholdIncomeHeatMap from '../common/assets/Local Data/census/median_income_data.js';
import { markerData } from '../common/assets/Local Data/MockMarkerData';
import { theme } from '../theme/theme';
import {
  FeatureProperty,
  HeatMapSelection,
  LayersComponentProps,
  LocationFeatures,
  MarkerOrHeatMap,
  MarkerSelection,
  SelectedMarker,
  SetDataSources,
  SetHeatMapSelection,
  SetMarkerSelection,
  SetSelectedMarker
} from './MapTypes';

// number of allowed selections, subject to change based on ui/ux and graph team suggestions
const ALLOWED_MARKERS = 2;
const ALLOWED_BOTH = 2;

// all of the local data we have available
// TODO: pull this from backend! need distinct split between marker & heat map
const heatMapData = [medianHouseholdIncomeHeatMap, kitchenFaciltiesHeatMap];
export const allData = flatten([markerData as any, heatMapData]);

const StyleBox = styled(Box)({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'left',
  '& > *': {
    margin: theme.spacing(1)
  },
  '& > * + *': {
    marginTop: theme.spacing(3)
  }
});

// this is how we show everything in options (disable none)
const showAll: MarkerOrHeatMap = [];

// handles change of selection
// ensures that popups will not stay when their markers disappear
export function handleChange(
  value: MarkerOrHeatMap,
  setMarkerSelection: SetMarkerSelection,
  setHeatMapSelection: SetHeatMapSelection,
  setSelectedMarker: SetSelectedMarker,
  selectedMarker: SelectedMarker,
  setDataSources: SetDataSources
) {
  const newMarkers: MarkerSelection[] = [];
  let newHeatMap: {} | HeatMapSelection = {};
  const allSelections: string[] = [];
  const allDataSources: string[] = [];
  // TODO: fix type errors here, I am unable to use the MarkerOrHeatMap type
  // eslint-disable-next-line
  value.forEach((table: MarkerSelection | HeatMapSelection) => {
    if (table.type === 'FeatureCollection') {
      const marker = table as MarkerSelection;
      newMarkers.push(marker);
      allDataSources.push(marker.source);
      marker.features.forEach((items: FeatureProperty[]) => {
        items.forEach((selection: FeatureProperty) => {
          allSelections.push(selection.properties.name);
        });
      });
    } else if (table.type === 'HeatMap') {
      const heatMap = table as HeatMapSelection;
      newHeatMap = heatMap;
      allDataSources.push(heatMap.source);
    }
  });
  setHeatMapSelection(newHeatMap);
  setMarkerSelection(newMarkers);
  setSelectedMarker(
    selectedMarker.filter(
      (obj: LocationFeatures) => obj.properties.name in allSelections
    )
  );
  const dataSourceDict: {
    key: number;
    label: string;
  }[] = [];
  allDataSources.forEach((source: string, i: number) => {
    dataSourceDict.push({ key: i, label: source });
  });
  setDataSources(dataSourceDict);
}

// handles disabling options, only two markers or one marker & one heat map allowed
export function handleDisable(
  // TODO: fix type errors here, I am unable to use the MarkerOrHeatMap type
  // eslint-disable-next-line
  allData: any[],
  markerSelection: MarkerSelection[],
  heatMapSelection: HeatMapSelection,
  // TODO: fix type errors here, I am unable to use the MarkerOrHeatMap type
  // eslint-disable-next-line
  option: any
) {
  // disable all options if max number of markers are already selected
  if (markerSelection.length === ALLOWED_MARKERS) {
    return allData.includes(option);
  }
  if (Object.keys(heatMapSelection).length) {
    // disable all options if max number of total selections are already made
    if (markerSelection.length + 1 === ALLOWED_BOTH) {
      return allData.includes(option);
    }
    // disable heat map options if a heat map is already selected
    return allData
      .filter(obj => obj.type !== 'FeatureCollection')
      .includes(option);
  }
  // don't disable any options if the user hasn't reached max markers or max selections
  return showAll.includes(option);
}

// this function creates the multi-seletion autocomplete component
export default function LayersComponent(props: LayersComponentProps) {
  const {
    markerSelection,
    setMarkerSelection,
    heatMapSelection,
    setHeatMapSelection,
    selectedMarker,
    setSelectedMarker,
    setDataSources
  } = props;
  return (
    <StyleBox>
      <Autocomplete
        multiple
        id="tags-outlined"
        options={allData}
        // TODO: make sure this handles data not existing once we are pulling from DB
        defaultValue={[allData[0], allData[2]]}
        // disables all options when the user has chosen more than the allowedSelections
        getOptionDisabled={option =>
          handleDisable(allData, markerSelection, heatMapSelection, option)
        }
        getOptionLabel={option => option.name}
        filterSelectedOptions
        // informs the layerSelection variable with the user's selection
        onChange={(event, value) =>
          handleChange(
            value,
            setMarkerSelection,
            setHeatMapSelection,
            setSelectedMarker,
            selectedMarker,
            setDataSources
          )
        }
        renderInput={params => (
          <TextField
            {...params}
            variant="standard"
            label="Layers"
            placeholder="Select up to Two Layers"
            fullWidth
          />
        )}
      />
    </StyleBox>
  );
}
