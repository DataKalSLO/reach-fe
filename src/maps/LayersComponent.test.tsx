import React from 'react';
import { configure, shallow } from 'enzyme';
import LayersComponent from './LayersComponent';
import Adapter from 'enzyme-adapter-react-16';

React.useLayoutEffect = React.useEffect;

configure({ adapter: new Adapter() });

describe('<LayersComponent/> with no props', () => {
  // supress warnings (mui & jest don't always get along)
  const originalConsoleError = console.error;
  beforeEach(() => {
    console.error = jest.fn(msg => {
      if (msg.includes('Warning: useLayoutEffect does nothing on the server')) {
        return null;
      } else {
        originalConsoleError(msg);
      }
    });
  });

  afterEach(() => {
    console.error = originalConsoleError;
  });
  const props = {
    markerSelection: [],
    setMarkerSelection: jest.fn().mockResolvedValue([]),
    heatMapSelection: [],
    setHeatMapSelection: jest.fn().mockResolvedValue([]),
    selectedMarker: [],
    setSelectedMarker: jest.fn().mockResolvedValue([]),
    dataSources: [],
    setDataSources: jest.fn().mockResolvedValue([])
  };
  const container = shallow(<LayersComponent {...props} />);
  it('should match with snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });
});
