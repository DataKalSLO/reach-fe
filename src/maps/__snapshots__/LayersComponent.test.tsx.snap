// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayersComponent/> with no props should match with snapshot 1`] = `"<div class=\\"MuiBox-root MuiBox-root-3 makeStyles-root-1\\"><div class=\\"MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon\\" role=\\"combobox\\" aria-expanded=\\"false\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink\\" data-shrink=\\"true\\" for=\\"tags-outlined\\" id=\\"tags-outlined-label\\">Layers</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd\\"><div role=\\"button\\" class=\\"MuiChip-root MuiAutocomplete-tag MuiChip-deletable\\" tabindex=\\"-1\\" data-tag-index=\\"0\\"><span class=\\"MuiChip-label\\">Colleges</span><svg class=\\"MuiSvgIcon-root MuiChip-deleteIcon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z\\"></path></svg></div><input type=\\"text\\" aria-invalid=\\"false\\" autoComplete=\\"off\\" id=\\"tags-outlined\\" placeholder=\\"Select up to Two Layers\\" value=\\"\\" class=\\"MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd\\" aria-autocomplete=\\"list\\" autoCapitalize=\\"none\\" spellcheck=\\"false\\"/><div class=\\"MuiAutocomplete-endAdornment\\"><button class=\\"MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty\\" tabindex=\\"-1\\" type=\\"button\\" aria-label=\\"Clear\\" title=\\"Clear\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeSmall\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path></svg></span></button><button class=\\"MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator\\" tabindex=\\"-1\\" type=\\"button\\" aria-label=\\"Open\\" title=\\"Open\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg></span></button></div></div></div></div></div>"`;
